accepted/future-releases/nnbd/number-operation-typing.md

```diff
@@ -62,47 +62,76 @@ T add<T extends num>(T a, T b) => a + b;
+Users also understand that `clamp` will return either the receiver or one of the arguments. If those all have the same type, the result will have that type.
 
 ## Solution
 
 ### Extended Rules
 
-We extend the special-casing rules of `+`, `-`, `*` and `%` to also cover calls of the `remainder` method, and to also work with type parameters which extend `num` , `int` or `double`. Finally, if the second operand is a `double` and the first is a `num`, the result is guaranteed to be a `double`.
+We extend the special-casing rules of `+`, `-`, `*` and `%` to also cover calls of the `remainder` method, and to also work with type parameters which extend `num` , `int` or `double`. Finally, if the second operand is a `double` and the first is a `num`, the result is guaranteed to be a `double`.
 
 That is:
 
-> For an expression `e` of one of the forms `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, where the static type of `e1` is a non-`Never` type *T* where *T* <: `num` and the static type of `e2` is a non-`Never` type *S* where *S* <: `num`:
+> For an expression `e` of one of the forms `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, where the static type of `e1` is a non-`Never` type *T* where *T* <: `num` and the static type of `e2` is a non-`Never` type *S* where *S* <: `num`:
 >
-> * If S <: T then the static type of `e` is *T*. 
-> * Otherwise If *T* <: *S* then the static type of `e` is *S*.
-> * Otherwise if *T* <: `int` and *S* <: `int` then the static type of `e` is `int`.
-> * Otherwise if *T* <: `double` or *S* <: *double* then the static type of *e* is `double`.
+> * If *T* <: `int` and *S* <: `int` then the static type of `e` is *T*.
+> * Otherwise if *T* <: `double` , then the static type of `e` is *T*.
+> * Otherwise if *S* <: `double` then the static type of `e` is `double`.
+> * Otherwise if *T* is not a subtype of `int` or `double`, then the static type of `e` is *T*.
 > * Otherwise the static type of *e* is `num`.
 
 We also special-case the `clamp` method:
 
 > For a normal invocation `e` of the form `e1.clamp(e2, e3)`, where the static types of `e1`, `e2` and `e3` are *T*<sub>1</sub>, *T*<sub>2</sub> and *T*<sub>3</sub> respectively, which are all non-`Never` subtypes of `num`:
 >
-> * If any of *T*<sub>1</sub>, *T*<sub>2</sub> or *T*<sub>3</sub> is a supertype of the other two, the first such is the static type of `e`.
+> * If any of *T*<sub>1</sub>, *T*<sub>2</sub> or *T*<sub>3</sub> is a supertype of the other two, the first such (in source order of the corresponding operands) is the static type of `e`.
 > * Otherwise, if all of *T*<sub>1</sub>, *T*<sub>2</sub> and *T*<sub>3</sub> are subtypes of `int`, the static type of `e` is `int`.
 > * Otherwise if all of *T*<sub>1</sub>, *T*<sub>2</sub> and *T*<sub>3</sub> are subtypes of `double`, the static type of `e` is `double`.
-> * Otherwise the static type of `e` is `num`.
+> * Otherwise the static type of `e` is `num`.
```

Comment:
It should be safe. Are we sure it will always be a subtype of `num`?

Reply:
The signature of `clamp` is `num clamp(num, num)` in all of `int`, `double`, `num`, so the invocation is an error unless *T2*, *T3* is assignable to `num`, and all these rules only apply if *T1* is a subtype of `num`.

So the interesting cases are when some of *T2*, *T3* are dynamic. In that case the actual arguments must still pass `arg as num`.

So it would probably be most helpful to say `min(num, LUB(T1, T2, T3))`, and that would be sound.

Reply:
What if the static type of *T2* and *T3* are `T extends Comparable<num>&double` and `T extends Comparable<num>&int` respectively. Is the LUB of that `Comparable<num>&num` or just `Comparable<num>`?

Reply:
How would that be possible? The bound of a type variable is declared, and there is no way to denote those `&` types.

Reply:
```dart
void foo<C extends Comparable<num>>(C x, C y) {
  num z = 0;
  if (x is int && y is double) {
    print([x, y].runtimeType);  // prints "List<Comparable<num>>"
    print([x, y, z].runtimeType);  // prints "List<Comparable<num>>"
    print([z.clamp(x, y)].runtimeType);  // prints "List<num>"
  }
}
main() {
  foo<Comparable<num>>(1, 1.5);
}
```

Here the static type of `x` is `C extends Comparable<num> & int`, and the static type of `y` is `C extends Comparable<num> & double`. The output suggests that the LIUB of those (and `num`) is actually `Comparable<num>`.

The static type of `z.clamp(x, y)` should still be `num`.
I'll handle this by saying that if the LUB is not <: `num`, the static type is `num`. Then the static type is still sound, even when the inputs are outrageous. (Won't happen in practice, but it should still be sound).

Reply:
Ah, the notation `C extends Comparable<num> & int` is new, we'd usually write `C & int`.

> I'll handle this by saying that if the LUB is not <: num, the static type is num.

Sounds good, that would be the same thing that I intended when I wrote `min(num, LUB(T1, T2, T3))`.